zerver/lib/send_email.py

```diff
@@ -416,23 +417,20 @@ def clear_scheduled_invitation_emails(email: str) -> None:
 
 
 @transaction.atomic(savepoint=False)
-def clear_scheduled_emails(user_ids: List[int], email_type: Optional[int] = None) -> None:
+def clear_scheduled_emails(user_id: int, email_type: Optional[int] = None) -> None:
     # We need to obtain a FOR UPDATE lock on the selected rows to keep a concurrent
     # execution of this function (or something else) from deleting them before we access
     # the .users attribute.
-    items = ScheduledEmail.objects.filter(users__in=user_ids).select_for_update()
+    items = (
+        ScheduledEmail.objects.filter(users__in=[user_id])
+        .prefetch_related("users")
+        .select_for_update()
+    )
     if email_type is not None:
         items = items.filter(type=email_type)
 
-    deduplicated_items = {}
     for item in items:
-        deduplicated_items[item.id] = item
-    for item in deduplicated_items.values():
-        # Now we want a FOR UPDATE lock on the item.users rows
-        # to prevent a concurrent transaction from mutating them
-        # simultaneously.
-        item.users.all().select_for_update()
```

Comment:
The new way of locking by using prefetch is definitely much better but I'm trying to understand whether there was a bug here? The reasoning here was that it's fine to lock .users a bit after SELECTing and locking the ScheduledEmail because we only want to delete the .users  - so if something concurrently mutated .users in the meantime, it doesn't matter because we're DELETEing these anyway

Reply:
Nope, there wasn't a bug here, just improving consistency.  I tried to explain this a bit in the commit message, but locking the item.users rows is not actually necessary at all, in some ways -- we only ever access them after locking the `item`, so we're guaranteed to already have serialized access.  But in case we ever _do_ start interacting with the `item` and `item.users` rows separately, we avoid a potential deadlock by taking them at-once, rather than in a specific order.

Reply:
Reworded the commit message to try to make this clearer.  Let me know if that helps!

Reply:
Yup, new phrasing is clear, thanks!